src/realm/util/sensitive_buffer.cpp

```diff
@@ -0,0 +1,230 @@
+        // it should succeed on the first try.
+        min_size *= 2;
+        max_size = std::max(4 * min_size, max_size);
+
+        // give up and don't try to ask the system to keep more than 90% memory resident in the process
+        if (max_size > 0.9 * limits->mem_size)
+            break;
+
+        BOOL ret = SetProcessWorkingSetSize(GetCurrentProcess(), min_size, max_size);
+        REALM_ASSERT_RELEASE_EX(ret != 0 && "SetProcessWorkingSetSizeEx", GetLastError());
+        limits->min_size = min_size;
+        limits->max_size = max_size;
+    }
+
+    // the loop above didn't succeed, provide some mem info and assert
+    MEMORYSTATUSEX mem;
+    mem.dwLength = sizeof(mem);
+    GlobalMemoryStatusEx(&mem);
+    REALM_ASSERT_RELEASE_EX(res != 0 && "VirtualLock()", err, mem.ullAvailPhys, mem.ullTotalPhys, mem.dwMemoryLoad,
+                            mem.ullAvailPageFile, mem.ullAvailVirtual);
+}
+
+SensitiveBufferBase::SensitiveBufferBase(size_t size)
+    : m_size(size_t(ceil((long double)size / CRYPTPROTECTMEMORY_BLOCK_SIZE) * CRYPTPROTECTMEMORY_BLOCK_SIZE))
+{
+    m_buffer = VirtualAlloc(nullptr, m_size, MEM_COMMIT | MEM_RESERVE, PAGE_READWRITE);
+    REALM_ASSERT_RELEASE_EX(m_buffer != NULL && "VirtualAlloc()", GetLastError());
+
+    lock_or_grow_working_size(m_buffer, m_size);
+}
+
+SensitiveBufferBase::~SensitiveBufferBase()
+{
+    if (!m_buffer)
+        return;
+
+    secure_erase(m_buffer, m_size);
+
+    BOOL ret = VirtualUnlock(m_buffer, m_size);
+    REALM_ASSERT_RELEASE_EX((ret != 0 || GetLastError() == ERROR_NOT_LOCKED) && "VirtualUnlock()", GetLastError());
+
+    ret = VirtualFree(m_buffer, 0, MEM_RELEASE);
+    REALM_ASSERT_RELEASE_EX(ret != 0 && "VirtualFree()", GetLastError());
+
+    m_buffer = nullptr;
+}
+
+void SensitiveBufferBase::protect() const
+{
+    // MEMO even though we try to lock the page with the buffer, and that should prevent it to be swapped,
+    //      locking is not reliable and may fail under high demand in the system due to some opaque reason,
+    //      which we can't recover from, so use second layer to protect the buffer if it is swapped
+    //      note: look at attempt_to_lock where it's expected
+    BOOL ret = CryptProtectMemory(m_buffer, DWORD(m_size), CRYPTPROTECTMEMORY_SAME_PROCESS);
+    REALM_ASSERT_RELEASE_EX(ret == TRUE && "CryptProtectMemory()", GetLastError());
+}
+
+void SensitiveBufferBase::unprotect() const
+{
+    BOOL ret = CryptUnprotectMemory(m_buffer, DWORD(m_size), CRYPTPROTECTMEMORY_SAME_PROCESS);
+    REALM_ASSERT_RELEASE_EX(ret == TRUE && "CryptUnprotectMemory()", GetLastError());
+}
+
+#else
+SensitiveBufferBase::SensitiveBufferBase(size_t size)
+    : m_size(size)
+{
+    m_buffer = mmap(nullptr, m_size, PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);
+    REALM_ASSERT_RELEASE_EX(m_buffer != MAP_FAILED && "mmap()", errno);
+
+    int ret;
+    ret = mlock(m_buffer, m_size);
```

Comment:
I anticipate we will run into problems when running this system call on actual devices. From what I can tell from the sparse documentation, android requires the root privileges. On iOS the recommended way is to do this is to use the secure enclave. Has this been tested on an actual iOS or android device?

Reply:
I am moderately surprised that it even compiles for iOS as it doesn't really make any sense there.

I don't think there's any reason for us to have non-Windows implementations of this feature at all. It's not a generally useful thing outside of the single customer's use-case, and they only care about Windows.

Reply:
The non-Windows path is just a copy of the Mongo’s server’s secure allocator - but I’m not sure it’s needed on mobile platforms where the system doesn’t even swap in the first place. Like Thomas said, the customer in question is only interested in Windows and that’s the only platform I did manual tests on.

Reply:
@ironage a cursory look at the kernel source seems to suggest that mlock needs the process to be privileged only if the process attempts to lock more memory than the configured value of RLIMIT_MEMLOCK: https://android.googlesource.com/kernel/common/+/refs/heads/android-mainline/mm/mlock.c. That might explain why Android Emulator tests weren’t failing, I believe. But in any case, mlock-ing might just be an overkill on a system that doesn’t swap in the first place.

Reply:
Which android emulator tests are you referring to? Core CI doesn't have that at the moment. :( 
Thanks for tracking down the sources. But agree that it doesn't seem suitable for mobile.

Reply:
Noted 😄